3_train/src/models.py

```diff
@@ -0,0 +1,351 @@
+        h_n, c_n = [], []
+
+        # expand bias vectors to batch size
+        bias_batch = (self.bias.unsqueeze(0).expand(batch_size, *self.bias.size()))
+
+        # calculate input gate only once because inputs are static
+        bias_s_batch = (self.bias_s.unsqueeze(0).expand(batch_size, *self.bias_s.size()))
+        i = torch.sigmoid(torch.addmm(bias_s_batch, x_s, self.weight_sh))
+
+        # perform forward steps over input sequence
+        for t in range(seq_len):
+            h_0, c_0 = h_x
+
+            # calculate gates
+            gates = (torch.addmm(bias_batch, h_0, self.weight_hh) +
+                     torch.mm(x_d[t], self.weight_ih))
+            f, o, g = gates.chunk(3, 1)
+
+            c_1 = torch.sigmoid(f) * c_0 + i * torch.tanh(g)
+            h_1 = torch.sigmoid(o) * torch.tanh(c_1)
+
+            # store intermediate hidden/cell state in list
+            h_n.append(h_1)
+            c_n.append(c_1)
+
+            h_x = (h_1, c_1)
+
+        h_n = torch.stack(h_n, 0)
+        c_n = torch.stack(c_n, 0)
+
+        if self.batch_first:
+            h_n = h_n.transpose(0, 1)
+            c_n = c_n.transpose(0, 1)
+
+        return h_n, c_n
+
+
+## From https://github.com/kratzert/ealstm_regional_modeling/blob/64a446e9012ecd601e0a9680246d3bbf3f002f6d/main.py
+
+class Model(nn.Module):
+    """Wrapper class that connects LSTM/EA-LSTM with fully connceted layer"""
+
+    def __init__(self,
+                 input_size_dyn: int,
+                 input_size_stat: int,
+                 hidden_size: int,
+                 output_size: int,
+                 initial_forget_bias: int = 5,
+                 dropout: float = 0.0,
+                 concat_static: bool = False,
+                 no_static: bool = False):
+        """Initialize model.
+
+        Parameters
+        ----------
+        input_size_dyn: int
+            Number of dynamic input features.
+        input_size_stat: int
+            Number of static input features (used in the EA-LSTM input gate).
+        hidden_size: int
+            Number of LSTM cells/hidden units.
+        output_size: int
+            Number of LSTM output units.
+        initial_forget_bias: int
+            Value of the initial forget gate bias. (default: 5)
+        dropout: float
+            Dropout probability in range(0,1). (default: 0.0)
+        concat_static: bool
+            If True, uses standard LSTM otherwise uses EA-LSTM
+        no_static: bool
+            If True, runs standard LSTM
+        """
+        super(Model, self).__init__()
+        self.input_size_dyn = input_size_dyn
+        self.input_size_stat = input_size_stat
+        self.hidden_size = hidden_size
+        self.initial_forget_bias = initial_forget_bias
+        self.dropout_rate = dropout
+        self.concat_static = concat_static
+        self.no_static = no_static
+
+        if self.concat_static or self.no_static:
```

Comment:
I like this, and more broadly, the class is a good minimal way to add the fully connected layer

Reply:
It makes it easy to compare the EA-LSTM and a vanilla LSTM with static and dynamic attributes concatenated together.